src/lib/es2015.promise.d.ts

```diff
@@ -1,3 +1,9 @@
+// The undefined case is for strictNullChecks false, in which case
+// undefined extends PromiseLike<infer U> is true, which would otherwise
+// make Awaited<undefined> -> unknown. When strictNullChecks is false
+// and T is undefined, should infer U be undefined, instead of unknown?
+type Awaited<T> = T extends undefined ? T : T extends PromiseLike<infer U> ? U : T;

```

Comment:
A problem with `Promise<Promise<number>>` is that it's equivalent to `Promise<number>`, but not assignable to it. Instead of making `Awaited` recursive, what do you think of [this attempt](https://github.com/microsoft/TypeScript/pull/33074) to resolve #27711, and avoid nested promises in the first place?

Reply:
It's still possible to write `Promise<Promise<number>>` and get the wrong experience, which is unfortunate. We've been investigating a [number](https://github.com/microsoft/TypeScript/pull/17077) of [ways](https://github.com/microsoft/TypeScript/pull/21613) to [resolve this](https://github.com/microsoft/TypeScript/pull/29317) for some time now. Until we have a firm decision on how to handle nested promises, I'm not sure I could approve this PR.

Reply:
What do you think of a solution like the one in #31117 in the mean time?